src/ng/compile.js

```diff
@@ -3836,8 +3836,17 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           return {
             pre: function ngPropPreLinkFn(scope, $element) {
               function applyPropValue() {
-                var propValue = ngPropGetter(scope);
-                $element.prop(propName, sanitizer(propValue));
+                var propValue = sanitizer(ngPropGetter(scope));
+
+                if (propValue !== undefined) {
+                  $element.prop(propName, propValue);
+                } else {
+                  if ($element.removeProp) {
+                    $element.removeProp(propName);
+                  } else {
+                    delete $element[0][propName];
```

Comment:
TBH, I would prefer setting it to undefined. `ngProp` is primarily useful for interacting with custom elements and I doubt deleting properties is a common usecase there.
I don't feel too strongly about it, though. Happy to go either way.

Reply:
@jbedard 
> I think it should be `delete` instead of assigning `undefined` because that's what jQuery does...

What do you mean by "what jQuery does"? Matching `removeProp`? I think the argument was we should neither `delete` nor use `removeProp` when available due to caveats listed at its API entry page:

> **Note:** Do not use this method to remove native properties such as checked, disabled, or selected. This will remove the property completely and, once removed, cannot be added again to element. Use `.prop()` to set these properties to `false` instead.

It doesn't seem to me using `removeProp` is a good choice here.

BTW: jQuery's `.prop(name, undefined)` does nothing while jqLite's version treats it as a getter. Resolving this inconsistency is covered in #11943 and won't fixed due to the feature freeze.

**NOTE:** in many cases setting the property to `undefined` will be a noop, e.g. in case of boolean props. It should be advised not to use `undefined` as a value unless that's expected. Or we could mirror jQuery's `prop` behavior here and for `undefined` values do nothing.

What do you think, @gkalpak & @jbedard?

Reply:
@gkalpak 
> Can you elaborate on that? How is it a noop?

You're right, my tests were faulty. :)

@jbedard 
> I thought we should use `removeProp` when available because of `jQuery.propFix`. Otherwise it's possible we will be removing a different property then we assigned...

Good point! By default jQuery defines `propFix` as:
```js
	propFix: {
		"for": "htmlFor",
		"class": "className"
	},
```
How does jqLite handle them, especially `for -> htmlFor`?

Another question is whether we want to support custom `jQuery.propFix` mappings but since this API is not a public one I think that's not necessary.

Reply:
There's also a few more prop fixes for casing: https://github.com/jquery/jquery/blob/3.3.1/src/attributes/prop.js#L128

Reply:
> `ngProp*` is primarily targetted towards custom elements which can have arbitrary property

Completely arbitrary? What about `class` vs `className`?

Reply:
I mean, custom elements can have any property (even `class`). For example, the following should set the element's text content to `Class: bar`:

```js
class MyFoo extends HTMLElement {
  set class(value) { this.textContent = `Class: ${value}`; }
}
customElements.define('my-foo', MyFoo);
...
const f = document.createElement('my-foo2');
f.class = 'bar';
```

Reply:
I see, you're allowed to use any property but if you don't define a setter in the component, native behavior will be used. For example:
```js
class MyFoo extends HTMLElement {}
customElements.define('my-foo', MyFoo);

const f = document.querySelector('my-foo');
f.className = 'bar';

console.log(f.getAttribute('class')); // 'bar'
```

Therefore, I think I agree with you. Let's set the property directly; if someone needs to use it on regular elements for attributes like `class` or `for` for any reason they can always use the full property name - which will be more consistent anyway.

Reply:
I agree direct property use is best if we are ok with that change. It could be considered a "breaking change" but I think that's ok at this point, especially since it's also a fix and the cases which will be "breaking" were broken already...

I'll make that change 👍

Reply:
Updated to just assign the property directly. This makes the diff much smaller, avoids the use of `delete`, makes jqLite/jQuery consistent, fixes the bug, etc...